common/migration/default-comparison.go

```diff
@@ -0,0 +1,157 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package migration
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+)
+
+var _ ComparisonFn[struct{}] = defaultComparisonFn[struct{}]
+
+// DefaultComparisonFn is a simple, suitable-for-low-throughput comparison function which
+// uses reflection to determine if the results match. Probably not good for high RPS endpoints
+// as it'll cause a significant amount of log noise. Can cause panics due to the use of the
+// cmp library (which panics for a number of conditions) so *not* for use without a recover.
+//
+// This function only compares structs/pointers to structs and errors. It will not attempt to compare
+// strings or primitive types and will just state them as not matching.
+func defaultComparisonFn[T any](
+	log log.Logger,
+	scope metrics.Scope,
+	activeRes T,
+	activeErr error,
+	backgroundRes T,
+	backgroundErr error) (isEqual bool) {
+
+	if activeErr != nil && backgroundErr != nil {
+		errDiff := cmp.Diff(activeErr, backgroundErr, cmpopts.EquateErrors())
```

Comment:
I'm not sure this will produce a result we want tbh - `EquateErrors()` does basically this:
```go
return activeErr.Is(backgroundErr) || backgroundErr.Is(activeErr)
```
which means they need to either be the same _instance_ or one needs to contain the other.

In some rare cases that'll be true, but usually I don't think it will be.  And if we add wrapping to one of those cases (e.g. both return `context.DeadlineExceeded`), it'll start to fail and probably mislead us.

Reply:
This was actually me just responding to a helpfully labelled panic by Diff about refusing to compare errors due to unexported fields and suggesting this. 

I'm wasn't really trying to solve for the errors.As/Is nested case since it seemed like overkill. Possibly I should just string match error strings for now even.